File: app/src/main/java/org/fossasia/susi/ai/adapters/viewholders/PieChartViewHolder.java
package org.fossasia.susi.ai.adapters.viewholders;

import android.graphics.Color;
import android.view.View;
import android.widget.LinearLayout;
import android.widget.TextView;

import com.github.mikephil.charting.charts.PieChart;
import com.github.mikephil.charting.data.Entry;
import com.github.mikephil.charting.data.PieData;
import com.github.mikephil.charting.data.PieDataSet;
import com.github.mikephil.charting.formatter.PercentFormatter;
import com.github.mikephil.charting.utils.ColorTemplate;

import org.fossasia.susi.ai.R;
import org.fossasia.susi.ai.model.ChatMessage;
import org.fossasia.susi.ai.rest.responses.susi.Datum;

import java.util.ArrayList;

import butterknife.BindView;
import butterknife.ButterKnife;
import io.realm.RealmList;

/**
 * <h1>Pie chart view holder</h1>
 */
public class PieChartViewHolder extends MessageViewHolder {

    @BindView(R.id.background_layout)
    public LinearLayout backgroundLayout;
    @BindView(R.id.text)
    public TextView chatTextView;
    @BindView(R.id.piechart)
    public PieChart pieChart;
    @BindView(R.id.timestamp)
    public TextView timeStamp;


    /**
     * Instantiates a new Pie chart view holder.
     *
     * @param view     the view
     * @param listener the listener
     */
    public PieChartViewHolder(View view, ClickListener listener) {
        super(view, listener);
        ButterKnife.bind(this, itemView);
    }

    /**
     * Inflate PieChart
     *
     * @param model the ChatMessage object
     */
    public void setView(ChatMessage model) {
        if (model != null) {
            try {
                chatTextView.setText(model.getContent());
                timeStamp.setText(model.getTimeStamp());
                pieChart.setUsePercentValues(true);
                pieChart.setDrawHoleEnabled(true);
                pieChart.setHoleRadius(7);
                pieChart.setTransparentCircleRadius(10);
                pieChart.setRotationEnabled(true);
                pieChart.setRotationAngle(0);
                pieChart.setDragDecelerationFrictionCoef(0.001f);
                pieChart.getLegend().setEnabled(false);
                pieChart.setDescription("");
                RealmList<Datum> datumList = model.getDatumRealmList();
                final ArrayList<Entry> yVals = new ArrayList<>();
                final ArrayList<String> xVals = new ArrayList<>();
                for (int i = 0; i < datumList.size(); i++) {
                    yVals.add(new Entry(datumList.get(i).getPercent(), i));
                    xVals.add(datumList.get(i).getPresident());
                }
                pieChart.setClickable(false);
                pieChart.setHighlightPerTapEnabled(false);
                PieDataSet dataSet = new PieDataSet(yVals, "");
                dataSet.setSliceSpace(3);
                dataSet.setSelectionShift(5);
                ArrayList<Integer> colors = new ArrayList<>();
                for (int c : ColorTemplate.VORDIPLOM_COLORS)
                    colors.add(c);
                for (int c : ColorTemplate.JOYFUL_COLORS)
                    colors.add(c);
                for (int c : ColorTemplate.COLORFUL_COLORS)
                    colors.add(c);
                for (int c : ColorTemplate.LIBERTY_COLORS)
                    colors.add(c);
                for (int c : ColorTemplate.PASTEL_COLORS)
                    colors.add(c);
                dataSet.setColors(colors);
                PieData data = new PieData(xVals, dataSet);
                data.setValueFormatter(new PercentFormatter());
                data.setValueTextSize(11f);
                data.setValueTextColor(Color.GRAY);
                pieChart.setData(data);
                pieChart.highlightValues(null);
                pieChart.invalidate();
            } catch (Exception e) {
                e.printStackTrace();
            }
        }
    }
}

File: app/src/main/java/org/fossasia/susi/ai/chat/adapters/viewholders/PieChartViewHolder.kt
package org.fossasia.susi.ai.chat.adapters.viewholders

import android.graphics.Color
import android.view.View
import android.widget.LinearLayout
import android.widget.TextView

import com.github.mikephil.charting.charts.PieChart
import com.github.mikephil.charting.components.Description
import com.github.mikephil.charting.data.PieData
import com.github.mikephil.charting.data.PieDataSet
import com.github.mikephil.charting.data.PieEntry
import com.github.mikephil.charting.formatter.PercentFormatter
import com.github.mikephil.charting.utils.ColorTemplate

import org.fossasia.susi.ai.R
import org.fossasia.susi.ai.data.model.ChatMessage

import java.util.ArrayList

import butterknife.ButterKnife
import kotterknife.bindView
import timber.log.Timber

class PieChartViewHolder(view: View, listener: MessageViewHolder.ClickListener) : MessageViewHolder(view, listener) {

    val backgroundLayout: LinearLayout by bindView(R.id.background_layout)
    val chatTextView: TextView by bindView(R.id.text)
    val pieChart: PieChart by bindView(R.id.piechart)
    val timeStamp: TextView by bindView(R.id.timestamp)

    init {
        ButterKnife.bind(this, itemView)
    }

    /**
     * Inflate PieChart
     *
     * @param model the ChatMessage object
     */
    fun setView(model: ChatMessage?) {
        if (model != null) {
            try {
                chatTextView.text = model.content
                timeStamp.text = model.timeStamp
                pieChart.setUsePercentValues(true)
                pieChart.isDrawHoleEnabled = true
                pieChart.holeRadius = 7f
                pieChart.transparentCircleRadius = 10f
                pieChart.isRotationEnabled = true
                pieChart.rotationAngle = 0f
                pieChart.dragDecelerationFrictionCoef = 0.001f
                pieChart.legend.isEnabled = false
                val description = Description()
                description.text = ""
                pieChart.description = description
                val datumList = model.datumRealmList
                val yVals = ArrayList<PieEntry>()
                val xVals = ArrayList<String>()

                if (datumList != null) {
                    for (item in datumList) {
                        yVals.add(PieEntry(item.percent, item.president))
                    }
                }
                pieChart.isClickable = false
                pieChart.isHighlightPerTapEnabled = false
                val dataSet = PieDataSet(yVals, "")
                dataSet.sliceSpace = 3f
                dataSet.selectionShift = 5f
                val colors = listOf(
                        ColorTemplate.COLORFUL_COLORS,
                        ColorTemplate.JOYFUL_COLORS,
                        ColorTemplate.VORDIPLOM_COLORS,
                        ColorTemplate.LIBERTY_COLORS,
                        ColorTemplate.PASTEL_COLORS)
                        .flatMap {
                            it.asIterable()
                        }
                dataSet.colors = colors
                val data = PieData(dataSet)
                data.setValueFormatter(PercentFormatter())
                data.setValueTextSize(11f)
                data.setValueTextColor(Color.GRAY)
                pieChart.data = data
                pieChart.highlightValues(null)
                pieChart.invalidate()
            } catch (e: Exception) {
                Timber.e(e)
            }
        }
    }
}
